torch/_subclasses/fake_tensor.py

```diff
@@ -853,6 +928,7 @@ def __init__(
         allow_non_fake_inputs=False,
         shape_env=None,
         static_shapes=None,
+        export=False,
```

Comment:
why do we need the export kwarg ? would be great to avoid if possible..

Reply:
The reason why I need to avoid doing this transformation during export is that when item_memo gets filled with a backed SymInt, it transforms deferred runtime asserts into guards. But while deferred runtime asserts are incorporated into the exported graph product, typically guards are ignored entirely.

Perhaps this is an export problem: we should always generate runtime assertions for guards in the graph, which I believe would eliminate the need for this. cc @angelayi @BoyuanFeng

Reply:
yea, transforming the guards here into runtime asserts makes sense to me

Reply:
This has some knock on effects, specifically, we will start generating runtime asserts for size guards that we didn't previously generate. I could split the difference in this PR by only generating runtime asserts for tensor inputs :think:

Reply:
I spent an honest attempt at this. I think it... could still possibly work? But it will be a bit invasive to export. I'm not keen on shoving it all into this patch. Here is my attempt: https://gist.github.com/ezyang/ad0f6da568b2c2b5c9427e0a5de53bde

The mechanical problem I'd gotten to before I gave up is that strict export is dropping the torch._check calls I'm inserting to enforce guards. Strict doesn't have this problem, because they manually call `insert_deferred_runtime_asserts` sufficiently late so the checks don't get removed. (AOTAutograd removes the torch._check calls in strict export, so we need to make sure we reinsert the deferred runtime asserts again.)

There's also some business going on in export where they are still generating asserts as assert_async on tensors (which obviously isn't going to propagate facts to ShapeEnv. I'm not entirely sure why the code is still doing it this way, this may need to get refactored.

At minimum these tests need fixing: https://gist.github.com/ezyang/d567607b415a9e76b3df038c447c6be1

Reply:
We discussed it at the export weekly and it looks like the export team has ongoing refactoring related to deferred runtime asserts that should potentially make it easier. So I propose that we land this PR as is, and then revisit export flag removal when those refactors land.

cc @pianpwk @avikchaudhuri @angelayi @tugsbayasgalan